### DISCUSSION TITLE
Bokeh in Jupyterlite

### DISCUSSION THREAD

**Discussion by MarcSkovMadsen** [bokeh/bokeh#11516] [General]
Jupyterlite is based on pyodide and runs entirely in the browser. It's already pretty amazing. I would like to use Bokeh and the HoloViz ecosystem in there.

I believe a lot of users would find Jupyterlite pretty amazing. A lot of the work I see out there consists of load data, transform using pandas, visualize, train a model or build an interactive data app. That is already possible if the viz is based on Matplotlib.

But I cannot import Bokeh using micropip.

https://jupyterlite.readthedocs.io/en/latest/_static/lab/index.html

![image](https://user-images.githubusercontent.com/42288570/120077848-b647d180-c0ac-11eb-8b7c-215d30761813.png)

## Solution

Solve what ever technical problems there may be and provide some examples in the docs.

**Comment by bryevdv** [comment]
dupe of https://github.com/bokeh/bokeh/issues/10572 (wheels will be available starting with 2.4)

**Comment by MarcSkovMadsen** [comment]
@bryevdv . My simple understanding is that wheels is one of the stepping stones. But probably some of the dependencies of Bokeh cannot be installed and that needs to be solved as well. And there are probably other issues. For example I can install plotly. But it does not work in there. Some of the code is not compatible with the jupyterlite environment.

My request is for Bokeh to work in Jupyterlite. And for that to be documented to users.

I'm writing it here because I see it as key for Bokeh to work in Jupyterlite. It's like before the iphone arrived. We would not understand how important it was for our apps to run on that platform.

**Comment by philippjfr** [comment]
I would love to see this too. That said:

>But probably some of the dependencies of Bokeh cannot be installed and that needs to be solved as well.

If the wheels solve it then great, if upstream dependencies prevent it from working then we should request compatibility in those libraries. It's certainly not a project anyone on the Bokeh team would commit to solving. Certainly don't object to keeping this issue open to track the progress towards compatibility though.

>I'm writing it here because I see it as key for Bokeh to work in Jupyterlite.

In the long run I also agree with this but in the short term I think exactly these kinds of compatibility issues will hinder adoption and I think it will take quite some time for this ecosystem to be anywhere near ready for production use cases.

Specifically for Bokeh the biggest obstacle I see is Tornado, specifically how the comms work in a setup where the server isn't actually a server. Exploring whether this is even possible will be the first and most complicated step.

**Comment by MarcSkovMadsen** [comment]
Totally agree with you @philippjfr . Right now this is a vision without a roadmap and resources. But it will be key longer term to be able to run in Jupyterlite and other pyodide  based environments. But maybe new wrappers of existing js libraries is what will quickly end up being used in there?

**Comment by philippjfr** [comment]
@bryevdv Any objection to keeping this open? If not I'm happy to move it to Panel to keep a compatibility checklist, since the concrete blocker in Bokeh itself (i.e. the wheels) have already been solved.

**Comment by bryevdv** [comment]
@philippjfr I leave it up to you

> I'm writing it here because I see it as key for Bokeh to work in Jupyterlite.

Just chiming in  my $0.02 which is that it is not at all clear to me that these current experiments will actually end up "being a thing"

**Comment by MarcSkovMadsen** [comment]
For future reference @philippjfr . C.f. https://github.com/jtpio/jupyterlite/discussions/112. 

![image](https://user-images.githubusercontent.com/42288570/120090857-1b7be100-c106-11eb-984b-29f9aae17162.png)

Would you mind to reopen this @philippjfr ? I and potentially other can use it as a placeholder to measure the interest and compile information for the time being.

**Comment by MarcSkovMadsen** [comment]
Vega support is coming to jupyterlite. https://github.com/jtpio/jupyterlite/pull/113

**Comment by MarcSkovMadsen** [comment]
Plotly support is coming to Jupyterlite. C.f. https://github.com/jtpio/jupyterlite/pull/129

**Comment by MarcSkovMadsen** [comment]
Ipyleaflet support added https://twitter.com/martinrenou/status/1405821341425717248?s=21

**Comment by bryevdv** [comment]
@MarcSkovMadsen what is your purpose with these updates? The are not useful or helpful in any way IMO

**Comment by MarcSkovMadsen** [comment]
The purpose is two fold

1. One day these might be usefull sources for understanding what it takes to become jupyterlite compatible.

2. It's a list of currently supported plotting tools. It shows whether or not jupyterlite starts becoming useful for data viz and whether there is general interest in supporting the platform.

**Comment by bryevdv** [comment]
> One day these might be usefull sources for understanding what it takes to become jupyterlite compatible.

According to your screenshot above this is already understood and the only roadblock is a "pyiodide-WASM-native build of libyaml". So I don't see how this is actually the case. 

> It shows whether or not jupyterlite starts becoming useful for data viz and whether there is general interest in supporting the platform.

I'm going to speak plainly and just say that it comes off as nagging.

**Comment by MarcSkovMadsen** [comment]
As stated earlier I really just wanted to 

![image](https://user-images.githubusercontent.com/42288570/122634523-e77d5580-d0de-11eb-8319-70be7affb47d.png)

I don't want to give any one the impression I am nagging. So I will reflect on how to avoid that in the future.

Thanks for letting me know.

**Comment by MarcSkovMadsen** [comment]
As stated earlier I really just wanted to 

> use it as a placeholder to measure the interest and compile information for the time being.

I don't want to give any one the impression I am nagging. So I will reflect on how to avoid that in the future.

Thanks for letting me know.

**Comment by bryevdv** [comment]
Just noting, since the above issue no longer exists (I guess jptio turned off all issues?) that I think it would be fairly easy to make yaml an optional rather than a hard dependency. Wheels are coming with 2.4 release. That just leaves Tornado, as @philippjfr I don't know what you do about that. You could still do lots with Bokeh w/o Tornado, but making Tornado optional is not a trivial undertaking. However if folks are talking about Panel apps or Dask + Dashboard then just making it optional is not a solution. There's:

* Port tornado into pyodide -- is anyone looking into this?
* Liberate the bokeh server protocol from Tornado
 
 I am actually very interested in the second thing, regardless of pyodide. But it's not something I will ever get done myself as a "free time after work" project. I don't see how any of these things get done without some dedicated support.

**Reply by bryevdv** [reply]
Also: converting to a GH discussion

**Comment by bryevdv** [comment]
Just noting that progress on this was made for upcoming 3.7 (see https://github.com/bokeh/bokeh/issues/13817 for details)
